### PR TITLE
Logs Panel: move log panel options and add sort order

### PR DESCRIPTION
**Pull request by gtk-grafana**
Moving Log panel options to the panel header, and adding logs direction:
image

WARNING: Breaking change for existing log wrap setting in local storage!

### CODE REVIEW

File: src/Components/ServiceScene/LogOptionsScene.tsx [hunk 1]
```diff
@@ -42,40 +59,85 @@ export class LogOptionsScene extends SceneObjectBase<LogOptionsState> {
...
     marginBottom: 0,
   }),
-  horizontalInlineSwitch: css({
-    padding: `0 4px 0 0`,
+  buttonGroupWrapper: css({
+    margin: 0,
+    alignItems: 'center',
+  }),
+  buttonGroup: css({
+    // padding: `0 4px 0 0`,
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/LogOptionsScene.tsx:141]: clean up

File: src/Components/Table/LogsHeaderActions.tsx [hunk 2]
```diff
@@ -33,12 +38,11 @@ export function LogsPanelHeaderActions(props: {
 
 const styles = {
   visualisationType: css({
-    display: 'flex',
-    flex: '1',
-    justifyContent: 'space-between',
-    marginTop: '8px',
+    // display: 'flex',
```
Comment (gtk-grafana) [inline src/Components/Table/LogsHeaderActions.tsx:41]: clean up

File: src/Components/ServiceScene/LogOptionsScene.tsx [hunk 3]
```diff
@@ -1,30 +1,47 @@
...
-  constructor(state?: Partial<LogOptionsState>) {
+  constructor(state: LogOptionsState) {
+    const logsSortOrderFromLocalStorage = getLogOption('sortOrder');
     super({
       ...state,
-      wrapLines: Boolean(getLogOption('wrapLines')),
+      sortOrder: (logsSortOrderFromLocalStorage
+        ? logsSortOrderFromLocalStorage
+        : LogsSortOrder.Descending) as LogsSortOrder,
+      wrapLogMessage: Boolean(getLogOption('wrapLogMessage')),
```
Comment (gtk-grafana) [inline src/Components/ServiceScene/LogOptionsScene.tsx:31]: This is the breaking change for line wrap, not really necessary, but it's nice to have the options here consistent with the types.

File: src/Components/ServiceScene/LogOptionsScene.tsx [hunk 4]
```diff
@@ -1,30 +1,47 @@
...
 
+/**
+ * The options rendered in the logs panel header
+ */
 export class LogOptionsScene extends SceneObjectBase<LogOptionsState> {
   static Component = LogOptionsRenderer;
 
-  constructor(state?: Partial<LogOptionsState>) {
+  constructor(state: LogOptionsState) {
+    const logsSortOrderFromLocalStorage = getLogOption('sortOrder');
```
Comment (svennergr) [inline src/Components/ServiceScene/LogOptionsScene.tsx:25]: We could change `getLogOption` to `getLogOption<T>(localStorageKey: string, defaultValue: T)` to return `defaultValue` if it's not found. WDYT?

File: src/Components/ServiceScene/LogOptionsScene.tsx (hunk 3 above)
Comment (svennergr) [reply src/Components/ServiceScene/LogOptionsScene.tsx:31]: I think it's fine to do.

File: tests/exploreServicesBreakDown.spec.ts [hunk 5]
```diff
@@ -3,6 +3,7 @@ import { ExplorePage, PlaywrightRequest } from './fixtures/explore';
 import { testIds } from '../src/services/testIds';
 import { mockEmptyQueryApiResponse } from './mocks/mockEmptyQueryApiResponse';
 import { LokiQuery } from '../src/services/lokiQuery';
+import exp = require('node:constants');
```
Comment (svennergr) [inline tests/exploreServicesBreakDown.spec.ts:6]:
```suggestion
```

File: src/Components/ServiceScene/LogOptionsScene.tsx [hunk 6]
```diff
@@ -1,30 +1,47 @@
...
+  handleWrapLinesChange = (type: 'wrap' | 'nowrap') => {
+    this.setState({ wrapLogMessage: type === 'wrap' });
+    setLogOption('wrapLogMessage', type === 'wrap');
+    this.getParentScene().setLogsVizOption({ wrapLogMessage: type === 'wrap' });
+  };
+
+  onChangeLogsSortOrder = (sortOrder: LogsSortOrder) => {
+    this.setState({ sortOrder: sortOrder });
+    setLogOption('sortOrder', sortOrder);
+    this.getParentScene().setLogsVizOption({ sortOrder: sortOrder });
```
Comment (svennergr) [inline src/Components/ServiceScene/LogOptionsScene.tsx:44]: I think we need to either call this method initially, or set it in the logs panel directly. If I change ordering to "Oldest first", everything works fine. Reloading the page sets the toggle to "Oldest first", but logs are sorted by "newest first".

**Review by svennergr — COMMENTED** [review]
Left a few comments, especially the sorting on page refresh needs a fix, but I really like the UX of it and using the RadioGroups for all options is really slick.

File: src/Components/ServiceScene/LogOptionsScene.tsx [hunk 7]
```diff
@@ -1,30 +1,47 @@
...
+      wrapLogMessage: Boolean(getLogOption('wrapLogMessage')),
     });
   }
 
-  handleWrapLinesChange = (e: ChangeEvent<HTMLInputElement>) => {
-    const checked = e.target.checked;
-    this.setState({ wrapLines: checked });
-    setLogOption('wrapLines', checked);
-    this.getParentScene().setLogsVizOption({ wrapLogMessage: checked });
+  handleWrapLinesChange = (type: 'wrap' | 'nowrap') => {
```
Comment (matyax) [inline src/Components/ServiceScene/LogOptionsScene.tsx:35]:
```suggestion
  handleWrapLinesChange = (wrap: boolean) => {
```
Also seems more compatible with `wrapLogMessage` which is also a boolean.

File: src/Components/ServiceScene/LogOptionsScene.tsx (hunk 4 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/LogOptionsScene.tsx]: Good call, added

File: src/Components/ServiceScene/LogOptionsScene.tsx (hunk 6 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/LogOptionsScene.tsx:41]: Fixed, I'll add this case to the e2e test as well

File: src/Components/ServiceScene/LogOptionsScene.tsx (hunk 7 above)
Comment (gtk-grafana) [reply src/Components/ServiceScene/LogOptionsScene.tsx]: Fixed!

**Review by svennergr — APPROVED** [review]
🚀 🚀 🚀 

I think this scrolling behavior is an issue with the logspanel itself, but worth calling out:

https://github.com/user-attachments/assets/f77849a5-9e97-4a82-9836-5865acab3d38

**Comment by gtk-grafana** [comment]
@svennergr yeah the panel tries to keep the same lines in the viewport for better or for worse. It's probably more annoying in a QA context then in a real life situation

**Comment by svennergr** [comment]
> @svennergr yeah the panel tries to keep the same lines in the viewport for better or for worse. It's probably more annoying in a QA context then in a real life situation

That's not the case in core Explore though, so somewhere is a disconnect, I think.

**Comment by gtk-grafana** [comment]
oh, I didn't realize that! @matyax is this a bug in Explore, or Explore logs?

**Comment by matyax** [comment]
I think this is the first time I see this behavior in the logs panel. Probably within LogsPanel or something about Scenes.

**Pull request merged**